packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1328 @@
+    }
+    assert(_debugTrackOrphans(noLongerOrphan: child));
+    dropChild(child);
+  }
+
+  List<RenderBox>? _debugOrphans;
+
+  // When a child is inserted into a slot currently occupied by another child,
+  // it becomes an orphan until it is either moved to another slot or removed.
+  bool _debugTrackOrphans({RenderBox? newOrphan, RenderBox? noLongerOrphan}) {
+    assert(() {
+      _debugOrphans ??= <RenderBox>[];
+      if (newOrphan != null) {
+        _debugOrphans!.add(newOrphan);
+      }
+      if (noLongerOrphan != null) {
+        _debugOrphans!.remove(noLongerOrphan);
+      }
+      return true;
+    }());
+    return true;
+  }
+
+  /// Throws an exception saying that the object does not support returning
+  /// intrinsic dimensions if, in debug mode, we are not in the
+  /// [RenderObject.debugCheckingIntrinsics] mode.
+  ///
+  /// This is used by [computeMinIntrinsicWidth] et al because viewports do not
+  /// generally support returning intrinsic dimensions. See the discussion at
+  /// [computeMinIntrinsicWidth].
+  @protected
+  bool debugThrowIfNotCheckingIntrinsics() {
+    assert(() {
+      if (!RenderObject.debugCheckingIntrinsics) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('$runtimeType does not support returning intrinsic dimensions.'),
+          ErrorDescription(
+            'Calculating the intrinsic dimensions would require instantiating every child of '
+            'the viewport, which defeats the point of viewports being lazy.',
+          ),
+        ]);
+      }
+      return true;
+    }());
+    return true;
+  }
+
+  @override
+  double computeMinIntrinsicWidth(double height) {
+    assert(debugThrowIfNotCheckingIntrinsics());
+    return 0.0;
+  }
+
+  @override
+  double computeMaxIntrinsicWidth(double height) {
+    assert(debugThrowIfNotCheckingIntrinsics());
+    return 0.0;
+  }
+
+  @override
+  double computeMinIntrinsicHeight(double width) {
+    assert(debugThrowIfNotCheckingIntrinsics());
+    return 0.0;
+  }
+
+  @override
+  double computeMaxIntrinsicHeight(double width) {
+    assert(debugThrowIfNotCheckingIntrinsics());
+    return 0.0;
+  }
+
+  @override
+  void applyPaintTransform(RenderBox child, Matrix4 transform) {
+    // Hit test logic relies on this always providing an invertible matrix.
+    parentDataOf(child).applyPaintTransform(transform);
+  }
+
+  @override
+  void dispose() {
+    _clipRectLayer.layer = null;
+    super.dispose();
+  }
+}
+
+/// A delegate used by [RenderTwoDimensionalViewport] to manage its children.
+///
+/// [RenderTwoDimensionalViewport] objects reify their children lazily to avoid
+/// spending resources on children that are not visible in the viewport. This
+/// delegate lets these objects create, reuse and remove children.
+abstract class TwoDimensionalChildManager {
+  void _startLayout();
+  void _buildChild(ChildVicinity index);
+  void _reuseChild(ChildVicinity index);
+  void _endLayout();
```

Comment:
Maybe consider making the methods public. That way, people can implement their own `TwoDimensionalChildManager` if they wanted to and having a public class with no visible methods is kinda strange.

Reply:
Yeah, I think that is where we started a conversation about finding a way to make it private.
Would folks actually be able to implement their own? So much of the implementation relies on private class members.